pkg/kubelet/nodestatus/setters.go

```diff
@@ -126,8 +126,8 @@ func NodeAddress(nodeIPs []net.IP, // typically Kubelet.nodeIPs
 
 				nodeIPTypes := make(map[v1.NodeAddressType]bool)
 				for _, nodeAddress := range cloudNodeAddresses {
-					if nodeAddress.Address == nodeIP.String() {
-						enforcedNodeAddresses = append(enforcedNodeAddresses, v1.NodeAddress{Type: nodeAddress.Type, Address: nodeAddress.Address})
+					if net.ParseIP(nodeAddress.Address).Equal(nodeIP) {
+						enforcedNodeAddresses = append(enforcedNodeAddresses, v1.NodeAddress{Type: nodeAddress.Type, Address: nodeIP.String()})
```

Comment:
I'm not familiar with the cloud providers implementations, can this cause any issue, that the cloud provider is setting the address in one format in `nodeAddress.Address` and we canonicalize it here?
should we keep `nodeAddress.Address`

Reply:
It's probably safer to leave this as nodeAddress.Address and use net.IP.Equal for comparing. 

It would be nice though if all addresses were parsed before setting the status here but it's too late to enforce that I think

Reply:
The problem is that ipv6 allows for same address to be represented in different formats. 
As long as both ips are sanitized, are represented by `net.IP` and have been proved to be the same by `net.ParseIP(nodeAddress.Address).Equal(nodeIP)` , it is safe to pick one of them. This PR is choosing to use the one already in `net.IP` format instead of continuing to use the ip represented by string.
Ip addresses(particularly ipv6) should not be compared as strings. There have been a few couple fixes in the past due to this. I think this change is safe from that perspective. 
Changing the datatype of `nodeAddress.Address` to `net.IP` would be the right fix per me, but im unsure the surface area of that change, since that is used across cloud provider contract.

Reply:
I don't think nobody disagrees on the format issue, but the problem is that it changes the current behavior and clients that use string comparison and have been working until now will hotloop since they are setting one IP and we are changing it ... despite they are equivalent ... is the same issue you are reporting but the other way around

1. SET node-ip 2001:0db8:0000:0000:0000:0000:0000:0000
2. GET node-ip <- 2001:0db8::
3. // huh, is not what I meant back to 1

Reply:
My perspective is that this change is not changing thh behavior. I wil take an example and explain. Please let me know if my structure of thinking is right.
the nodeips are coming from 2 places. [1] `node-ip` provided to kubelet [ref](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L1114) [2] nodeip provided by cloud provider [ref](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/nodestatus/setters.go#L143) .

 
[1] is a parsed address and is always becomes a shorthand notation, i.e `2600:1f14:1d4:d101:0:0:0:ba3d` becomes  `2600:1f14:1d4:d101::ba3d`.
[2] is a ip represented as a [string](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L5145) coming from cloud provider as `nodeAddress.Address`.  Now there are 3 scenarios to think about. The `nodeAddress.Address` is 
 - a full octet: eg. `FE80:CD00:0CDE:0CDE:1257:0CDE:211E:729C`
   In this case the behavior stays unchanged with the PR
   `nodeAddress.Address` and `nodeIP.String()` have the same values.
 - shorthand ipv6: eg: `2600:1f14:1d4:d101::ba3d`
   In this case the behavior stays unchanged with the PR. `nodeAddress.Address` and `nodeIP.String()` have the same values.
 - fully expanded ipv6: eg `2600:1f14:1d4:d101:0:0:0:ba3d`
   `nodeAddress.Address` and `nodeIP.String()` have the same values do have different values and the scenario does not work anyway. We're fixing that in the PR/


Can you help me identify which situation will cause behavior change? I'm also ok with keeping `nodeAddress.Address` in the node address , but wanted to take a last try to understand what case won't work. If you insist on keeping `nodeAddress.Address` , i will make the change.

Reply:
> fully expanded ipv6: eg 2600:1f14:1d4:d101:0:0:0:ba3d
nodeAddress.Address and nodeIP.String() have different values and the scenario does not work anyway. We're fixing that in the PR/

> An issue in aws cloud provider is not sanitizing the imds node ip before being sent to kubelet. As a result, the address provided to the node-ip in kubelet and the one received from cloud provider mismatches.

I see 2 issues: 

The main issue here is that kubelet does string comparison internally between node.Address and node.IP, and you fixed it.

The other issue is about mutating the value, because this is a string field, and we can't assume clients will do the conversion to IPs, so if I set `2600:1f14:1d4:d101:0:0:0:ba3d` it should be legit to think that I want to get that value back, otherwise I should send the compressed value in the first hand. 

If we mutate the node.Address you'll break clients that set `2600:1f14:1d4:d101:0:0:0:ba3d` and expect the same value, despite they are the same because they represent the same IP address. (However, I do agree that we should have canonicalized the IP addresses on the API, it will save us from a lot of headaches like https://github.com/kubernetes/kubernetes/pull/104368, but that trains seems long gone)

But maybe we should get confirmation from apimachinery group, I may be wrong on my assumptions.
ping @liggitt

Reply:
> and we can't assume clients will do the conversion to IPs, so if I set 2600:1f14:1d4:d101:0:0:0:ba3d it should be legit to think that I want to get that value back

My perspective is that it would be certainly a new use case to set `2600:1f14:1d4:d101:0:0:0:ba3d` and expect it back. As long as we're not breaking exitsting clients, we should have autonomy to define the behavior, which we are right now, by saying that node ip will always be sanitized for ipv6. For ipv4 ips, the changes and the comments don't apply.

Reply:
> My perspective is that it would be certainly a new use case to set 2600:1f14:1d4:d101:0:0:0:ba3d and expect it back(since it doesn't work today
it works if node-ip is set to :: or 0.0.0.0 , doesn't it?

Also, this code has a lot of branches, in case that we want to canonicalize we should do it in more places ...

I

Reply:
> it works if node-ip is set to :: or 0.0.0.0 , doesn't it?

Yes, and this change won't affect it. ipv4 is not impacted by the canonicalization(actually i don't know. can it?)
`::` keeps working as expected because it has no `0` in it. However `0::0` today doesn't work and will start working after this change. Of course the node ip in this newly working case is `::` even when `0::0` is passed from cloud provider.

Reply:
what I say is that if you set address to ` 2600:1f14:1d4:d101:0:0:0:ba3d` and `node-ip` is set to `::` , the node.Status.Address will be set to 2600:1f14:1d4:d101:0:0:0:ba3d`, it will not fail

Reply:
I understand your point now and I agree with this usecase and that there will be inconsistency in behavior when node-ip is `::`.
I'll make the change to use `nodeAddress.Address` instead of `nodeIP.String()` . Thanks for taking the time to educate me.

Reply:
done. The change is now just the if condition